python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+    pass
+```
+
+Notes:
+
+* See discussion in "Runtime - Workflow Stepped Execution" on whether to reuse `asyncio.sleep`
+* Like `temporal.activity.wait*` we choose the "wait_" prefix here as Python often does, e.g. for events which this is
+  most analogous to
+
+#### `temporalio.workflow` Miscellaneous
+
+```python
+
+@dataclass(frozen=True)
+class Info:
+    pass
+
+def is_replaying() -> bool:
+    pass
+
+def info() -> Info:
+    pass
+
+def now() -> datetime:
+    pass
+
+logger = LoggerAdapter(logging.getLogger(__name__), None)
+```
+
+Notes:
+
+* Fields of `Info` not shown, but notably absent is `is_replaying` so it can remain immutable
+  * Instead, like Java and Go, we move that to the package level
+* Should we move `is_replaying` to a place that makes it more clear it's unsafe?
+* It's clearer to require uses of `temporalio.workflow.now()` than to wait for and require overriding `datetime.now()`
+  when a deterministic sandbox appears. It is clearer to disallow deterministic calls than implicitly replace some.
+* `logger` uses a special adapter that is skipped during replay
+
+#### `temporalio.workflow.CancellationScope`
+
+Rough sketch:
+
+```python
+class CancellationScope:
+    @property
+    @staticmethod
+    def current() -> CancellationScope:
+        pass
+
+    def __init__(
+        self,
+        *,
+        # Default is current
+        parent: Optional[CancellationScope] = None,
+        detached: bool = False,
+        timeout: Optional[timedelta] = None,
+    ) -> None:
+        pass
+
+    def _push(self) -> None:
+        pass
+
+    def _pop(self) -> None:
+        pass
+
+    await def run(self, fn: Callable[..., Union[Awaitable[T], T]]) -> T:
+        pass
+
+    def __enter__(self) -> CancellationScope:
+        pass
+
+    def __exit__(self) -> None:
+        pass
+
+    def cancel(self) -> None:
+        pass
+
+    @property
+    def cancelled(self) -> bool:
+        pass
+```
+
+Notes:
+
+* Unlike TypeScript, Python still allows you to explicitly cancel individual activities and child workflows if you
+  started them with `start_`
+  * The consistency of having everything that has a `start_` be cancellable has value. I still expect most will use the
+    `execute_` form though
+* The workflow will have a root scope
+* This will work using `with CancellationScope()` and similar and is powered by `contextvars`
+  * `run` is async because the callable may have an async result, but `with` is the preferred approach for explicit use
+* `start_activity`/`execute_activity`/`start_child_workflow`/`execute_child_workflow` will implicitly use the current
+  cancellation scope
+* Do we need extra capabilities such as:
+  * Ability to explicitly add cancellables?
+  * Ability to unset the current instead of just forcing a detached one each time you want to opt out?
+* Should we disallow the `parent` to even be set in the constructor?
+  * Thinking we probably shouldn't allow you to set that
```

Comment:
I've never seen this used in TS (or any instantiation of CancellationScopes as a matter of fact)

Reply:
It seems harmless to have and provides some clarity. Technically I made up this use case for TS:

```typescript
parent = CancellationScope.current()
scope1 = new CancellationScope()
scope2 = new CancellationScope({ parent })
f1 = scope1.run(activityA())
f2 = scope1.run(activityB())
f3 = scope2.run(activityC())
f4 = scope1.run(activityD())
f5 = scope2.run(activityE())
```

But I admit that's rare.